src/ray/gcs/gcs_server/gcs_resource_report_poller.cc

```diff
@@ -0,0 +1,116 @@
+#include "ray/gcs/gcs_server/gcs_resource_report_poller.h"
+
+namespace ray {
+namespace gcs {
+
+GcsResourceReportPoller::GcsResourceReportPoller(
+    uint64_t max_concurrent_pulls,
+    std::shared_ptr<GcsResourceManager> gcs_resource_manager,
+    std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool)
+    : max_concurrent_pulls_(max_concurrent_pulls),
+      gcs_resource_manager_(gcs_resource_manager),
+      raylet_client_pool_(raylet_client_pool),
+      poll_period_ms_(boost::posix_time::milliseconds(
+          RayConfig::instance().gcs_resource_report_poll_period_ms())),
+      poll_timer_(polling_service_) {}
+
+void GcsResourceReportPoller::Start() {
+  polling_thread_ = std::unique_ptr<std::thread>(new std::thread{[&]() {
+    RAY_LOG(ERROR) << "Polling thread created";
+    polling_service_.post([&]() { Tick(); });
+    polling_service_.run();
+  }});
+}
+
+void GcsResourceReportPoller::Tick() {
+  GetAllResourceUsage();
+
+}
+
+void GcsResourceReportPoller::GetAllResourceUsage() {
+  RAY_LOG(ERROR) << "Getting all resource usage";
+
+  {
+    absl::MutexLock guard(&mutex_);
+    for (const auto &pair : nodes_to_poll_) {
+      poll_state_.to_pull.push_back(pair.second);
+    }
+  }
+}
+
+void GcsResourceReportPoller::LaunchPulls() {
+  absl::MutexLock guard(&mutex_);
+  while (!poll_state_.to_pull.empty() && poll_state_.inflight_pulls.size() < max_concurrent_pulls_) {
+    rpc::Address &address = poll_state_.to_pull.back();
+    NodeID node_id = NodeID::FromBinary(address.raylet_id());
+
+    auto raylet_client = raylet_client_pool_->GetOrConnectByAddress(address);
+    raylet_client->RequestResourceReport(
+                                         [this, node_id](const Status &status, const rpc::RequestResourceReportReply &reply) {
+                                           // TODO (Alex): This callback is always posted onto the main thread. Since most
```

Comment:
If asio schedules in FIFO order this will result in a bounded delay for everything else on the main thread, but we sure verify asio schedules the way we expect to.